src/core/manager.c

```diff
@@ -3736,6 +3818,72 @@ int manager_deserialize(Manager *m, FILE *f, FDSet *fds) {
         return manager_deserialize_units(m, f, fds);
 }
 
+int manager_pin_all_cgroup_bpf_programs(Manager *m) {
+        int r;
+        Unit *u;
+
+        assert(m);
+
+        HASHMAP_FOREACH(u, m->units) {
+                size_t i;
+                BPFProgram *p;
+                BPFProgram *builtin_progs[] = {u->bpf_device_control_installed,
+                                           u->ip_bpf_ingress,
+                                           u->ip_bpf_ingress_installed,
+                                           u->ip_bpf_egress,
+                                           u->ip_bpf_egress_installed};
+                Set *custom_progs[] = {u->ip_bpf_custom_ingress,
+                                       u->ip_bpf_custom_ingress_installed,
+                                       u->ip_bpf_custom_egress,
+                                       u->ip_bpf_custom_egress_installed};
+
+                for (i = 0; i < ELEMENTSOF(builtin_progs); i++) {
+                        p = builtin_progs[i];
+                        if (!p)
+                                continue;
+
+                        r = set_ensure_put(&m->bpf_limbo_progs, NULL, p);
+                        if (r < 0) {
+                                log_error_errno(r, "Cannot store BPF program for reload: %m");
+                                continue;
+                        }
+
+                        (void) bpf_program_ref(p);
+                }
+
+                for (i = 0; i < ELEMENTSOF(custom_progs); i++)
+                        SET_FOREACH(p, custom_progs[i]) {
+                                r = set_ensure_put(&m->bpf_limbo_progs, NULL, p);
+                                if (r < 0) {
+                                        log_error_errno(r, "Cannot store BPF program for reload: %m");
+                                        continue;
+                                }
+
+                                (void) bpf_program_ref(p);
+                        }
+        }
+
+        log_debug("Pinned %d BPF programs", set_size(m->bpf_limbo_progs));
```

Comment:
Same here. `log_unit_debug()`.

Reply:
At this point there's no unit any more, since we're out of the loop. I can add it inside the loop, but that's going to be a lot of logging inside a pretty tight loop. For now I'll keep it like this, unless you have any objections :-)

Reply:
Ouch. Indeed.